File: doc/codes/python/examples/tutorial/output-intensity.py
import asyncio

import numpy as np

import autd3_link_ethercrab as ethercrab
import autd3_modulation as modulation
import autd3_pattern as pattern
from autd3 import Client, ClientConfig
from autd3.commands import Modulation, Pattern, SetSilencer
from autd3.geometry import Autd3, Geometry
from autd3.units import Hz, m, s
from autd3.value import Intensity, SamplingConfig

async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await Client.open(
        geometry,
        ethercrab.EtherCrabLinkOption(),
        ClientConfig(),
    )

    target = geometry.center() + np.array([0.0, 0.0, 150.0])
    wavelength = pattern.wavelength(340 * m / s)

    # ANCHOR: pattern_intensity
    patterns = geometry.pattern_buffer()
    pattern.focus(
        geometry,
        target,
        wavelength,
        pattern.FocusOption(intensity=Intensity(0x80)),
        patterns,
    )
    # ANCHOR_END: pattern_intensity

    # ANCHOR: modulation
    mod_buf = modulation.modulation_buffer()
    modulation.sine(
        200.0 * Hz,
        modulation.SineOption(
            amplitude=0xFF,
            offset=0x80,
            sampling_config=SamplingConfig.FREQ_4K,
        ),
        mod_buf,
    )
    # ANCHOR_END: modulation

    builder = client.datagram_builder()
    builder.push(SetSilencer())
    builder.push(Pattern(patterns))
    builder.push(Modulation(SamplingConfig.FREQ_4K, mod_buf))
    for frame in builder.build():
        await client.send_checked(frame)

    await client.stop()
    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/output-intensity.rs
use anyhow::Result;

use autd3_rs::commands::{Modulation, Pattern, SetSilencer};
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{Hz, m, mm, s};
use autd3_rs::value::{Intensity, SamplingConfig};
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let target = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);

    // ANCHOR: pattern_intensity
    let mut patterns = geometry.pattern_buffer();
    autd3_rs_pattern::focus(
        &geometry,
        target,
        wavelength,
        &autd3_rs_pattern::FocusOption {
            intensity: Intensity(0x80),
            ..Default::default()
        },
        &mut patterns,
    );
    // ANCHOR_END: pattern_intensity

    // ANCHOR: modulation
    let mut modulation = autd3_rs_modulation::modulation_buffer();
    autd3_rs_modulation::sine(
        200 * Hz,
        &autd3_rs_modulation::SineOption {
            amplitude: 0xFF,
            offset: 0x80,
            sampling_config: SamplingConfig::FREQ_4K,
            ..Default::default()
        },
        &mut modulation,
    )?;
    // ANCHOR_END: modulation

    let mut builder = client.datagram_builder();
    builder
        .push(SetSilencer::default())
        .push(Pattern::new(&patterns))
        .push(Modulation::new(SamplingConfig::FREQ_4K, &modulation));
    for frame in &builder.build()? {
        client.send_checked(frame).await?;
    }

    client.stop().await?;
    client.close().await?;
    Ok(())
}
